Controlnet training (#2545)
huggingface/diffusers: examples/controlnet/README.md

## Installing the dependencies

Before running the scripts, make sure to install the library's training dependencies:

**Important**

To make sure you can successfully run the latest versions of the example scripts, we highly recommend **installing from source** and keeping the install up to date as we update the example scripts frequently and install some example-specific requirements. To do this, execute the following steps in a new virtual environment:
```bash
git clone https://github.com/huggingface/diffusers
cd diffusers
pip install -e .
```

Then cd in the example folder and run
```bash
pip install -r requirements.txt
```

And initialize an [🤗Accelerate](https://github.com/huggingface/accelerate/) environment with:

```bash
accelerate config
```

Or for a default accelerate configuration without answering questions about your environment

```bash
accelerate config default
```

Or if your environment doesn't support an interactive shell e.g. a notebook

```python
from accelerate.utils import write_basic_config
write_basic_config()
```

> ## Circle filling dataset

Author: Instructions need to be updated for new dataset.

Reviewer: +1 to that. 

We should add a section to guide our users on that. Cc: @apolinario @patrickvonplaten @williamberman

> # Controlnet training example

Reviewer: ```suggestion
# ControlNet training example
```

[Adding Conditional Control to Text-to-Image Diffusion Models](https://arxiv.org/abs/2302.05543) by Lvmin Zhang and Maneesh Agrawala.

> Using the pretrained models we can provide control images (for example, a depth map) to control Stable Diffusion text-to-image generation so that it follows the structure of the depth image and fills in the details. `train_controlnet.py` script shows how to train a controlnet from scratch with simple example data.

Reviewer: ```suggestion
Using the pretrained models, we can provide control images (for example, a depth map) to control Stable Diffusion text-to-image generation so that it follows the structure of the control image and steers the generated image accordingly. `train_controlnet.py` script shows how to train a ControlNet from scratch with a simple dataset.
```

Reviewer 2: I just removed the paragraph. We have other docs that explain what controlnet is

> This example is based on the [training example in the original Controlnet repository](https://github.com/lllyasviel/ControlNet/blob/main/docs/train.md) and it trains a controlnet that can fill circles using a small synthetic dataset.

Reviewer: ```suggestion
This example is based on the [training example from the original ControlNet repository](https://github.com/lllyasviel/ControlNet/blob/main/docs/train.md) and it trains a ControlNet that can fill circles using a small synthetic dataset.
```

Reviewer: Maybe also link the dataset we'll use?

## Installing the dependencies

Before running the scripts, make sure to install the library's training dependencies:

**Important**

To make sure you can successfully run the latest versions of the example scripts, we highly recommend **installing from source** and keeping the install up to date as we update the example scripts frequently and install some example-specific requirements. To do this, execute the following steps in a new virtual environment:
```bash
git clone https://github.com/huggingface/diffusers
cd diffusers
pip install -e .
```

Then cd in the example folder and run
```bash
pip install -r requirements.txt
```

And initialize an [🤗Accelerate](https://github.com/huggingface/accelerate/) environment with:

```bash
accelerate config
```

Or for a default accelerate configuration without answering questions about your environment

```bash
accelerate config default
```

Or if your environment doesn't support an interactive shell e.g. a notebook

```python
from accelerate.utils import write_basic_config
write_basic_config()
```

## Circle filling dataset

> The original dataset is hosted in the [controlnet repo](https://huggingface.co/lllyasviel/ControlNet/blob/main/training/fill50k.zip). We re-uploaded it to be compatible with datasets [here](https://huggingface.co/datasets/fusing/fill50k). Note that datasets handles dataloading

Reviewer: Let's maybe try to use "ControlNet".

Reviewer: ```suggestion
he original dataset is hosted in the [ControlNet repo](https://huggingface.co/lllyasviel/ControlNet/blob/main/training/fill50k.zip). We re-uploaded it to be compatible with `datasets` [here](https://huggingface.co/datasets/fusing/fill50k). Note that `datasets` handles dataloading within the training script.
```

> within the training script.

Reviewer: ```suggestion
```

> Our training examples use stable diffusion 1.5 as the original set of controlnet models were trained from it. However, controlnet can be trained to augment any stable diffusion compatible model.

Reviewer: ```suggestion
Our training examples use [Stable Diffusion 1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5) as the original set of ControlNet models were trained from it. However, ControlNet can be trained to augment any Stable Diffusion compatible model (such as [CompVis/stable-diffusion-v1-4](https://huggingface.co/CompVis/stable-diffusion-v1-4)).
```

## Training

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=4
```

This default configuration requires ~38GB VRAM.

> Gradient accumulation with a smaller batch size can be used to reduce training requirements to ~20 GB VRAM

Reviewer: ```suggestion
Gradient accumulation with a smaller batch size can be used to reduce training requirements to ~20 GB VRAM.
```

What happens when using `xformers`? Maybe we should just add a note like:

> We haven't rigorously test all the optimization techniques supported by `diffusers` for training but we think it should be possible to further reduce the memory footprint. This can be done by using mixed-precision, memory-efficient attention with `xformers`, and, 8-bit Adam for training.

WDYT?

 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=1 \
 --gradient_accumulation_steps=4 \
 --gradient_checkpointing \
 --use_8bit_adam
```

## Training on a 12 GB GPU

Optimizations:
- Gradient checkpointing
- bitsandbyte's 8-bit optimizer
- xformers
- set grads to none

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=1 \
 --gradient_accumulation_steps=4 \
 --gradient_checkpointing \
 --use_8bit_adam \
 --enable_xformers_memory_efficient_attention \
 --set_grads_to_none
```

> 

Reviewer: Really like the progressive disclosure of the optimizations to make it more and more accessible for the consumer GPUs. 

I think we should add a note here:

```suggestion
 When using `enable_xformers_memory_efficient_attention`, please make sure to install `xformers` by `pip install xformers`.

  offload_optimizer_device: cpu
  offload_param_device: cpu
  zero3_init_flag: false
  zero_stage: 2
distributed_type: DEEPSPEED
```

See [documentation](https://huggingface.co/docs/accelerate/usage_guides/deepspeed) for more DeepSpeed configuration options.

Changing the default Adam optimizer to DeepSpeed's Adam
`deepspeed.ops.adam.DeepSpeedCPUAdam` gives a substantial speedup but
it requires CUDA toolchain with the same version as pytorch. 8-bit optimizer
does not seem to be compatible with DeepSpeed at the moment.

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=1 \
 --gradient_accumulation_steps=4 \
 --gradient_checkpointing \
 --enable_xformers_memory_efficient_attention \
 --set_grads_to_none \
 --mixed_precision fp16
```

> ## Running the trained controlnet

Reviewer: ```suggestion
## Running the trained ControlNet
```

> The trained model can be run same as the original controlnet pipeline with the newly trained controlnet.

Reviewer: ```suggestion
The trained model can be run the same as the original ControlNet pipeline with the newly trained ControlNet.
```

Set `base_model_path` and `controlnet_path` to the values `--pretrained_model_name_or_path` and 
`--output_dir` were respectively set to in the training script.

```py
from diffusers import StableDiffusionControlNetPipeline, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import torch

base_model_path = "path to model"

> controlnet_path = "path to controlnet"

Reviewer: IMO it's much better to provide code snippets here that just work out of the box. For example this:

https://github.com/huggingface/diffusers/tree/main/examples/text_to_image#inference

Reviewer 2: We do this in the dreambooth docs as well. Better to force the user to change a name if it's going to need to be changed regardless

> # ControlNet training example

Reviewer 3: Nice doc page!

Reviewer 3: Could you maybe add a W&B link here?

Reviewer 2: https://github.com/huggingface/diffusers/pull/2545/commits/9e87526c760ed16e8020ee9a3a954286a170e2ae

```bash
git clone https://github.com/huggingface/diffusers
cd diffusers
pip install -e .
```

Then cd in the example folder and run
```bash
pip install -r requirements.txt
```

And initialize an [🤗Accelerate](https://github.com/huggingface/accelerate/) environment with:

```bash
accelerate config
```

Or for a default accelerate configuration without answering questions about your environment

```bash
accelerate config default
```

Or if your environment doesn't support an interactive shell e.g. a notebook

```python
from accelerate.utils import write_basic_config
write_basic_config()
```

## Circle filling dataset

The original dataset is hosted in the [ControlNet repo](https://huggingface.co/lllyasviel/ControlNet/blob/main/training/fill50k.zip). We re-uploaded it to be compatible with `datasets` [here](https://huggingface.co/datasets/fusing/fill50k). Note that `datasets` handles dataloading within the training script.

> Our training examples use [Stable Diffusion 1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5) as the original set of ControlNet models were trained from it. However, ControlNet can be trained to augment any Stable Diffusion compatible model (such as [CompVis/stable-diffusion-v1-4](https://huggingface.co/CompVis/stable-diffusion-v1-4)).

Reviewer 3: ```suggestion
Our training examples use [Stable Diffusion 1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5) as the original set of ControlNet models were trained from it. However, ControlNet can be trained to augment any Stable Diffusion compatible model (such as [CompVis/stable-diffusion-v1-4](https://huggingface.co/CompVis/stable-diffusion-v1-4)) or [stabilityai/stable-diffusion-2-1](https://huggingface.co/stabilityai/stable-diffusion-2-1)
```

## Training

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=4
```

> This default configuration requires ~38GB VRAM.

Reviewer 3: We need this much? Because of the batch size? Can't we reduce this using xformers and gradient_accumulation? We should be able to get this down to 24GB VRAM I think

Reviewer 3: Ah should have kept on reading :sweat_smile: Let's maybe just advertise the example below requiring only 20GB of VRAM

 --train_batch_size=1 \
 --gradient_accumulation_steps=4
```

## Example results

#### After 300 steps with batch size 8

| |  | 
|-------------------|:-------------------------:|
| | red circle with blue background  | 
![conditioning image](./images/conditioning_image_1.png) | ![red circle with blue background](./images/red_circle_with_blue_background_300_steps.png) |
| | cyan circle with brown floral background | 
![conditioning image](./images/conditioning_image_2.png) | ![cyan circle with brown floral background](./images/cyan_circle_with_brown_floral_background_300_steps.png) |

#### After 6000 steps with batch size 8:

| |  | 
|-------------------|:-------------------------:|
| | red circle with blue background  | 
![conditioning image](./images/conditioning_image_1.png) | ![red circle with blue background](./images/red_circle_with_blue_background_6000_steps.png) |
| | cyan circle with brown floral background | 
![conditioning image](./images/conditioning_image_2.png) | ![cyan circle with brown floral background](./images/cyan_circle_with_brown_floral_background_6000_steps.png) |

> ## Training on a 16 GB GPU

Reviewer 3: The following is very nice!

Optimizations:
- Gradient checkpointing
- bitsandbyte's 8-bit optimizer

[bitandbytes install instructions](https://github.com/TimDettmers/bitsandbytes#requirements--installation).

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \
 --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \
 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=1 \
 --gradient_accumulation_steps=4 \
 --gradient_checkpointing \
 --use_8bit_adam
```

## Training on a 12 GB GPU

Optimizations:
- Gradient checkpointing
- bitsandbyte's 8-bit optimizer
- xformers
- set grads to none

```bash
export MODEL_DIR="runwayml/stable-diffusion-v1-5"
export OUTPUT_DIR="path to save model"

accelerate launch train_controlnet.py \
 --pretrained_model_name_or_path=$MODEL_DIR \
 --output_dir=$OUTPUT_DIR \
 --dataset_name=fusing/fill50k \
 --resolution=512 \
 --learning_rate=1e-5 \

>  --validation_image "./images/conditioning_image_1.png" "./images/conditioning_image_2.png" \

Reviewer 3: Can we maybe instead upload those to a dataset and then load it from there? I don't want the github repo to become to heavy - we shouldn't store images in the github repo IMO

Reviewer 2: Ah I didn't think about the git implications. Yes, that makes sense

Reviewer 2: https://github.com/huggingface/diffusers/pull/2545/commits/6fd13ea87d71b4a2f1147543d19cc1663b729890 done!

 --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
 --train_batch_size=1 \
 --gradient_accumulation_steps=4 \
 --gradient_checkpointing \
 --use_8bit_adam \
 --enable_xformers_memory_efficient_attention \
 --set_grads_to_none
```

When using `enable_xformers_memory_efficient_attention`, please make sure to install `xformers` by `pip install xformers`. 

## Training on an 8 GB GPU

We have not exhaustively tested DeepSpeed support for ControlNet. While the configuration does
save memory, we have not confirmed the configuration to train successfully. You will very likely

> have to make changes to the config to have a successful training run.

Reviewer 2: My deepspeed training run didn't give good results. I would like to leave the configuration in the documentation but with this caveat

Reviewer 3: ok for me!
